library/alloc/src/string.rs

```diff
@@ -1076,8 +1077,11 @@ impl String {
     #[must_use]
     #[stable(feature = "string_as_str", since = "1.7.0")]
     #[cfg_attr(not(test), rustc_diagnostic_item = "string_as_str")]
-    pub fn as_str(&self) -> &str {
-        self
+    #[rustc_const_unstable(feature = "const_vec_string_slice", issue = "129041")]
+    pub const fn as_str(&self) -> &str {
+        // SAFETY: String contents are stipulated to be valid UTF-8, invalid contents are an error
+        // at construction.
+        unsafe { str::from_utf8_unchecked(self.vec.as_slice()) }
```

Comment:
Is it correct that the `from_utf8_unchecked` is only needed for const-compatibility, and this should be reverted to its safe previous implementation when const-eval becomes more capable?

That should always be noted with a comment in the code. We even have a standard form for that, `FIXME(const-hack)`.

Reply:
Not specifically &mdash; this used to live in `Deref` and `DerefMut`, so the previous implementation just invoked the same `from_utf8_unchecked` via implicit deref.

I did invert the call structure (moved the primary impl here rather than in `Deref*`, which now are implemented using these methods) for const-compat reasons, but the `from_utf8_unchecked` would need to live somewhere, even if we had `const_trait`.

Reply:
Ah, fair -- thanks.